### PR TITLE
main

### PR DESCRIPTION
**Pull request by alexsolan1**

### CODE REVIEW

File: main.cc [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    
+    String(const char* data) {
+        m_len = strlen(data);
+        m_capacity = calc_reserve(m_len);
+        m_data = new char[m_capacity];
+        strncpy(m_data, data, m_len);
+        m_data[m_len] = 0;
+    }
+    
+    String(const String& s) {
```
Comment (alexsolan1) [inline main.cc:43]: fdsfdsdsfds

File: main.cc [hunk 2]
```diff
@@ -0,0 +1,210 @@
+#include <iostream>
+#include <string.h>
+#include <cassert>
+
+using namespace std;
+
+class String {
```
Comment (iwrinc) [inline main.cc:7]: But how the user would use this class? What file should be included?

File: main.cc [hunk 3]
```diff
@@ -0,0 +1,210 @@
+#include <iostream>
+#include <string.h>
```
Comment (iwrinc) [inline main.cc]: The standard says that whenever we need to include a C header xxx.h in C++, we have to `#include <cxxx>`

File: main.cc [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    String(const String& s) :
+        String(s.data(), s.length())
+    { }
+    
+    String(const char* data, size_t length) {
+        alloc_and_copy(data, length);
+    }
+    
+    String(unsigned capacity) {
+        m_data = new char[capacity];
```
Comment (iwrinc) [inline main.cc]: Consider *initializing* the members, rather than assigning them.

File: main.cc [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        m_capacity = capacity;
+        m_length = 1;
```
Comment (iwrinc) [inline main.cc]: Why 1? What does m_length denote?

File: main.cc [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+        m_length = s.length();
+        s.clean();
+    }
+        
+    ~String() {
+        delete[] m_data;
+    }
+    
+    void reserve(unsigned capacity) {
+        if (capacity < m_capacity) { return; }
```
Comment (iwrinc) [inline main.cc]: `<=`?

File: main.cc [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+    
+    unsigned int length() const {
+        return m_length;
+    }
+
+    unsigned int capacity() const {
+        return m_capacity;
+    }
+
+    const char* data() const {
```
Comment (iwrinc) [inline main.cc:73]: The assignment doesn't say that expicitly, but real std::string::data() allows data mutation.

File: main.cc [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    }
+ 
+    char& operator[] (unsigned int i) const {
+        assert(i < length());
+        return m_data[i];
+    }
+    
+    String& operator+=(const String& s) {
+        unsigned newlen = length() + s.length();
+        if (newlen + 1 >= m_capacity) {
```
Comment (iwrinc) [inline main.cc]: `>`?

File: main.cc [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+
+    String(const String& s) :
+        String(s.data(), s.length())
+    { }
+    
+    String(const char* data, size_t length) {
+        alloc_and_copy(data, length);
+    }
+    
+    String(unsigned capacity) {
```
Comment (iwrinc) [inline main.cc]: Although it was not stated in the assignment explicitly, we'd like to mimic the standard container behavior. If a standard container's constructor get an integer, it is the desired *size* (not the capacity).   In order to set capacity, containers define `reserve` member function.

File: main.cc [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+        m_capacity = length + 1;
+        m_length = length;
+        m_data = tmp;
+        strncpy(m_data, data, length);
+        m_data[m_length] = 0;
+    }
+    
+    void reallocate(unsigned length) {
+        String tmp(length);
+        tmp.m_length = m_length;
```
Comment (iwrinc) [inline main.cc:129]: It looks like the above constructor is not "good enough", so you have to adjust some members after it. Note that after you change the semantics of the ctor, as described above, the things might look better here.

File: main.cc [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+        tmp.add(*this, 0);
+        swap(*this, tmp);
+    }
+    
+    void add(const String& s, unsigned location) {
```
Comment (iwrinc) [inline main.cc]: Very dangerous function. Consider calling it something like `unsafe_add`.

File: main.cc [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+        return m_data;
+    }
+     
+    char* end() {
+        return m_data + length();
+    }
+   
+private:
+    friend void swap(String& first, String& second) {
+        swap(first.m_capacity, second.m_capacity);
```
Comment (iwrinc) [inline main.cc:113]: In this particular case all the members are built-in types. But what if some of them were a user-defined type with it's own `swap`?  
Note that `swap` is ADL-based. So the common idiom looks like this:
using std::swap;
swap(left, right);

File: main.cc [hunk 13]
```diff
@@ -0,0 +1,210 @@
+#include <iostream>
+#include <string.h>
+#include <cassert>
+
+using namespace std;
```
Comment (iwrinc) [inline main.cc:5]: Not in a namepsace/global scope, please!

**Review by iwrinc — COMMENTED** [review]
Great! It looks much better now!

File: main.cc (hunk 2 above)
Comment (alexsolan1) [reply main.cc:7]: I know I should separate the .h and the .cc, but I don't have a normal environment at home, so I work with the online compiler.

File: main.cc (hunk 9 above)
Comment (alexsolan1) [reply main.cc]: But in this case, if it is the size and not the capacity, I am left with un-initialized data?

File: main.cc (hunk 5 above)
Comment (alexsolan1) [reply main.cc]: Right. Should be 0. Will fix.

File: main.cc (hunk 6 above)
Comment (alexsolan1) [reply main.cc]: Right, of course.

File: main.cc (hunk 8 above)
Comment (alexsolan1) [reply main.cc]: Right again.

File: main.cc (hunk 12 above)
Comment (alexsolan1) [reply main.cc:113]: OK. Fixed.

File: main.cc (hunk 11 above)
Comment (alexsolan1) [reply main.cc]: Indeed (But it is private). I can change the name and add an assertion that capacity suffices.

File: main.cc (hunk 3 above)
Comment (alexsolan1) [reply main.cc]: Great. I didn't know. Now I know :) Fixed.

File: main.cc (hunk 13 above)
Comment (alexsolan1) [reply main.cc:5]: In the next exercises it is allowed... so I allowed myself in advance :) I promise that it just for this exercise...

File: main.cc (hunk 4 above)
Comment (alexsolan1) [reply main.cc]: Ok, will do.

File: main.cc (hunk 10 above)
Comment (alexsolan1) [reply main.cc:129]: But - I use the reallocate in the reserve, so I need it to adjust the capacity. Actually the naming here is not good. I will change it.

File: oneliners.cc [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+#include <numeric>
+#include <vector>
+#include <sstream>
+
+using namespace std;
+
+int main()
+{
+    vector<int> v(100);
+    iota(v.begin(), v.end(), 0);
```
Comment (iwrinc) [inline oneliners.cc:12]: But it's 2 lines (if I got it right, and it should be solution for (1) ).

File: one_liners2.cc [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+#include <vector>
+#include <string>
+#include <boost/algorithm/string/classification.hpp>
+
+using namespace std;
+
+// 1. Xor to each char
+string xored_string(string s, char n) {
+    return accumulate(s.begin(), s.end(), string(""), [n](
+        const string& acc, const char& c){return acc + char(c^n);});
```
Comment (iwrinc) [inline one_liners2.cc:13]: `const char& c` ?! :-)

File: one_liners2.cc [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+}
+// reverse
+string my_reverse(string s) {
```
Comment (iwrinc) [inline one_liners2.cc]: `const string&`

File: one_liners2.cc [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    return accumulate(s.crbegin(), s.crend(), string(""), [](
+        const string& acc, const char& c){return acc + c;});
```
Comment (iwrinc) [inline one_liners2.cc]: Formally, that's a correct solution. But could you please think of a much more simple (and potentially fast) one?

File: one_liners2.cc [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+// check all capitals
+bool all_capitals(const vector<string>& names) {
+    return accumulate(names.begin(), names.end(), true, [](
+        bool acc, const string& name){
+            return acc &
```
Comment (iwrinc) [inline one_liners2.cc]: You probably mean `&&`  (BTW, what would be the difference?)

File: one_liners2.cc [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+                    boost::is_upper()
+                );});
+}           
+
+// flatten vector of vectors of int
+vector<int> flat_vec(const vector<vector<int>>& vectors) {
+    return accumulate(vectors.begin(), vectors.end(), vector<int>{},
+        [](vector<int> acc, const vector<int>& cur) {
+            move(cur.begin(), cur.end(), back_inserter(acc));
+            return acc;
```
Comment (iwrinc) [inline one_liners2.cc:38]: More than 1 line.

Comment (alexsolan1) [reply one_liners2.cc:38]: It was specifically allowed to do 2 for this one.... :)

File: one_liners2.cc (hunk 18 above)
Comment (alexsolan1) [reply one_liners2.cc]: yeah, it is logical 'and' not bitwise.  Does the compiler converts the booleans to numeric to perform the  bitwise &?
Practically, I believe that although it is weird, it should work, as long as it is mapped true->1 and false->0.
I will fix, of course.

File: one_liners2.cc (hunk 16 above)
Comment (alexsolan1) [reply one_liners2.cc]: Right

File: one_liners2.cc (hunk 17 above)
Comment (alexsolan1) [reply one_liners2.cc]: Actually, there is std::reverse, which is a one-liner and I am sure that it works faster than my one-liner, but it is probably not the purpose of this ex?

File: oneliners.cc (hunk 14 above)
Comment (alexsolan1) [reply oneliners.cc:12]: Hmm, right. Is something like 
vector<int> v{1,3,5,7,45} is an answer? or should we do something that can be extended to long vectors?

File: main.cc (hunk 9 above)
Comment (iwrinc) [reply main.cc]: You'll have _value-initialized_ data (please, read about value-initialization).

File: oneliners.cc (hunk 14 above)
Comment (iwrinc) [reply oneliners.cc:12]: Well, the assignment does not say you have push _different_ elements ;-)

File: one_liners2.cc (hunk 17 above)
Comment (iwrinc) [reply one_liners2.cc]: I didn't mean this function, of course...
I guess the whole point of this exercise is to learn the rich (and some say bloated) interface of the standard library. So, just take a look at `std::string` interface :)

File: one_liners2.cc (hunk 18 above)
Comment (iwrinc) [reply one_liners2.cc]: Yes, in this case _integral promotion_ takes place (see C++20 standard, 7.3.6).  The problem is, however, that bitwise operations are _fully_ evaluated, while the logical ones have _short-circuit_ evaluation. So this is a nice example of how a single symbol might significantly affect performance :).

File: one_liners2.cc (hunk 19 above)
Comment (iwrinc) [reply one_liners2.cc:38]: oh, you're right!

File: oneliners.cc (hunk 14 above)
Comment (alexsolan1) [reply oneliners.cc:12]: Yes, but if I should use it as a base for all the other operations, it will be a bad test case to have like
vector v{100, 5}...
Of course I could do it and then create a more interesting one, to test the other operations :)

File: optional.h [hunk 20]
```diff
@@ -0,0 +1,31 @@
+#include <cassert>
+#include <iostream>
+#include <string>
+
+template <class T>
+class Optional {
```
Comment (iwrinc) [inline optional.h:15]: Please, test `Optional` with the following class:
```
struct my_class {
    my_class(int) {}
};
```

File: optional.h [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+	Optional() = default;
+	Optional(const T& t) :
+		m_value(t),
+		m_valid(true)
+	{ }
+	operator bool() const {
+		return m_valid;
+	}
+
+	T& get() {
```
Comment (iwrinc) [inline optional.h:40]: So, what would happen if the user attempts to access "empty" optional (built in release mode)? What is the desired behavior?

File: one_liners2.cc (hunk 18 above)
Comment (alexsolan1) [reply one_liners2.cc]: Hmmm,  I see!

File: optional.h (hunk 20 above)
Comment (alexsolan1) [reply optional.h:15]: It doesn't work since my_class is not std::is_default_constructible, I guess I should add the requirement to the definition of Optional.

File: optional.h (hunk 21 above)
Comment (alexsolan1) [reply optional.h:40]: In release mode in my implementation he will have an empty object.
I could also throw an exception.
But I still have the feeling that I am doing something wrong here, since if T is a big object even when empty, we (I) allocate it on the stack even for non valid Optional.

File: optional.h (hunk 20 above)
Comment (alexsolan1) [reply optional.h:15]: I can also implement my optional with variant, in this case it does not need to have T as default_constructible. But  I am not sure it is fair:)

Comment (alexsolan1) [reply optional.h:15]: And I could also allocate memory on the stack for T, keep inside a pointer of T and when I need to create an object of T I will use this memory. Thus also  I won't need a default_constructible type.

Comment (iwrinc) [reply optional.h:15]: If the optional requires a default-constructible T, why couldn't we just use more simple and less obfuscated std::pair<bool, T>?
Of course it should not have such a constraint.
Besides, the assignment requires not to use free store ("heap").

File: optional.h (hunk 21 above)
Comment (iwrinc) [reply optional.h:40]: Please, read the assignment thoroughly.

File: optional.h (hunk 20 above)
Comment (alexsolan1) [reply optional.h:15]: But using pair<bool, T> was exactly my naive implementation (just didn't use explicitly the pair)!
I understand that we want it work also with non default-constructible and therefore my solution is not good enough.
I didn't suggest to allocate memory on the heap, because the ex explicitly mentioned it.
I suggest to allocate memory on the stack (char buf[sizeof(T)])  and when we actually use the T call the new expression with this buffer
T* tptr = new(buf) T(t).

**Comment by alexsolan1** [comment]
Something is weird with your last comments - I can't comment on each separately. But:
1) I added alignment
2) I added ctor with Forwarding (no additional copy)
3) I added throw exception (indeed it was written in the assignment and I missed it, sorry....)

File: optional.h [hunk 22]
```diff
@@ -12,8 +20,17 @@ class Optional {
 	Optional(const T& t) :
```
Comment (iwrinc) [inline optional.h:20]: In what context is this ctor called now?

File: optional.h [hunk 23]
```diff
@@ -23,17 +40,21 @@ class Optional {
...
 
 	T& get() {
-		assert(m_valid);
+		if (!m_valid) {
+			throw BadOptional();
+		}
 		return *m_value;
 	}
 		
 	const T& get() const {
```
Comment (iwrinc) [inline optional.h:47]: Cool! But now we have 2 identical functions (copy&paste). Could we do some code-reuse? (I.e., implement one function and call it from the other one)

File: optional.h [hunk 24]
```diff
@@ -23,17 +40,21 @@ class Optional {
...
-		assert(m_valid);
+		if (!m_valid) {
+			throw BadOptional();
+		}
 		return *m_value;
 	}
 	
 private:
-	char buf[sizeof(T)];
+	alignas(alignof(T)) char buf[sizeof(T)];
```
Comment (iwrinc) [inline optional.h:55]: Great! This is the right solution.
(Just note that we have `std::aligned_storage` exactly for this purpose.)

**Review by iwrinc — APPROVED** [review]

**Review by iwrinc — APPROVED** [review]

File: derive.h [hunk 25]
```diff
@@ -45,28 +49,32 @@ ostream& operator<<(ostream& os, const Expression& ex) {
 
 class CoefExpression : public Expression {
 public:
-	CoefExpression(double a_coef): coef(a_coef) 
+	CoefExpression(double coef): m_coef(coef) 
 	{ }
 	
-	virtual bool is_zero() { return coef == 0.; }
+	virtual bool is_zero() { return coef() == 0.; }
```
Comment (iwrinc) [inline derive.h]: Could it be `const`?

File: derive.h [hunk 26]
```diff
@@ -45,28 +49,32 @@ ostream& operator<<(ostream& os, const Expression& ex) {
...
 	virtual ostream& printTo(ostream& os) const {
-		os << coef;
+		os << coef();
 		return os;
 	}
-	double coef;
+	double coef() const { return m_coef;}
+	friend class BinaryExpression;
+protected:
+	virtual optional<char> opt_var() const { return optional<char>{}; }
```
Comment (iwrinc) [inline derive.h]: We have `std::nullopt` for this purpose.

File: derive.h [hunk 27]
```diff
@@ -45,28 +49,32 @@ ostream& operator<<(ostream& os, const Expression& ex) {
...
+	double m_coef;
 };
 
 
 class XExpression : public Expression {
 public:
 	virtual ostream& printTo(ostream& os) const {
```
Comment (iwrinc) [inline derive.h]: Prefer `override`.

File: derive.h [hunk 28]
```diff
@@ -45,28 +49,32 @@ ostream& operator<<(ostream& os, const Expression& ex) {
...
 class CoefExpression : public Expression {
 public:
-	CoefExpression(double a_coef): coef(a_coef) 
+	CoefExpression(double coef): m_coef(coef) 
 	{ }
 	
-	virtual bool is_zero() { return coef == 0.; }
+	virtual bool is_zero() { return coef() == 0.; }
 	virtual shared_ptr<Expression> derivative() const {
 		return std::dynamic_pointer_cast<Expression>(
```
Comment (iwrinc) [inline derive.h]: Why is dynamic (or any) cast needed here?

File: derive.h [hunk 29]
```diff
@@ -0,0 +1,291 @@
...
+			new CoefExpression{0.}
+		); 
+	}
+
+	return make_shared_expression(
+		new BinaryExpression{lhs, rhs, mult}
+	);
+}
+
+shared_ptr<Expression> pow(shared_ptr<Expression> lhs, shared_ptr<Expression> rhs) {
```
Comment (iwrinc) [inline derive.h]: Note that passing `shared_ptr` by value causes refcount increment, which involves expensive atomic operations.

File: derive.h [hunk 30]
```diff
@@ -0,0 +1,291 @@
...
+	if (rhs->is_zero()) {
+		return lhs;
+	}
+	
+	return make_shared_expression(
+		new BinaryExpression{lhs, rhs, plus}
+	);
+}
+
+shared_ptr<Expression> subtract(shared_ptr<Expression> lhs, shared_ptr<Expression> rhs) {
```
Comment (iwrinc) [inline derive.h]: Looks like almost copy-paste from `add`, doesn't it?

File: derive.cc [hunk 31]
```diff
@@ -0,0 +1,24 @@
...
+	cout << "h=" << h << "derivative: " << h.derivative() << endl;
+	Func long_func = 2. * h + 5. * pow((f + 20.), 2.);
+	cout << "long_func = " << long_func << "derivative: " << long_func.derivative() << endl;
+	Func g = f*0.;
+	cout << "g(x)=" << g << " g'(x) " << g.derivative() << endl;	
+	Func f_mult_g = f * g;
+	cout << "h is: " << h << " derivative is: " << h.derivative() << endl;
+	cout << "f_mult_g is: " << f_mult_g << " derivative is: " << f_mult_g.derivative() << endl;
+	Func sub_func = f * 3. - x;
+	cout << "sub_func(x) = " << sub_func << " sub_func'(x) = " << sub_func.derivative() << endl;
```
Comment (iwrinc) [inline derive.cc:24]: Could you please add tests with zero coeff and very small coeff.

**Review by iwrinc — COMMENTED** [review]
General note: although this is legitimate design, it has some significant drawbacks  (I'd call it "javaish" design).
1) It uses enums and dynamic type queries, instead of C++ type system and compile-time checks. This leads to performance loss (dynamic casts are very slow, as they are usually implemented via string comparison) and poor maintainability (think of adding a new operation or expression type to all the numerous switch cases and dynamic type queries...).
2) It utilizes multiple virtual functions, which also might affect performance.

After you finish this solution, I'd recommend to try and propose an alternative design, which takes much more advantage of C++ compiler. Hint: templates and `boost::variant` :).
Such a solution would have its own advantages and drawbacks, and I believe it would be nice to compare them.

File: my_variant.h [hunk 32]
```diff
@@ -0,0 +1,118 @@
...
+	}
+	// need to add emplacement ctors...
+
+	template <class U>
+	my_variant(U&& u) :
+		m_index(get_index<U, T, types...>::value),	
+ 		m_member(new (&m_aligned_storage) U{u} )
+	{ }
+	~my_variant() {
+		destroy_all<decltype(*this), T, types...>::f(*this);
```
Comment (iwrinc) [inline my_variant.h]: You solved the destructor problem "ad hoc", which is fine :). But actually, it was just a symptom of the more general problem: what if you (or the user!) had to apply some other function to the contained value (e.g. print() etc)? In your case the only such function was the destructor, that's why you encountered this problem here.
So, the general problem is the following: we need a *generic* mechanism that would allow us to apply any function to the *contained* type. When you have such a mechanism, you're able to invoke it in the destructor as well! (Hint: usually we call such a mechanism "visitor".)

File: my_variant.h [hunk 33]
```diff
@@ -0,0 +1,118 @@
...
+	}
+	my_variant(my_variant<T, types...>&&) {
+		assert(sizeof(T) == 0 && "Not implemented");
+	}
+	// need to add emplacement ctors...
+
+	template <class U>
+	my_variant(U&& u) :
+		m_index(get_index<U, T, types...>::value),	
+ 		m_member(new (&m_aligned_storage) U{u} )
```
Comment (iwrinc) [inline my_variant.h]: Looks like `&&` is unused...

File: my_variant.h [hunk 34]
```diff
@@ -0,0 +1,118 @@
...
+    std::integral_constant<std::size_t, 1 + get_index<T, Ts...>::value> {};
+
+template <typename T>
+struct get_index<T> {
+	static_assert(sizeof(T) == 0, "Element is not found");
+};
+
+template <class Obj, class... types> struct destroy_all;
+template <class T, class... types>
+struct my_variant {
```
Comment (iwrinc) [inline my_variant.h:57]: Missing assignment (copy/move)

File: my_variant.h [hunk 35]
```diff
@@ -0,0 +1,118 @@
...
+
+template <typename T, typename... Ts>
+struct get_index<T, T, Ts...> : std::integral_constant<std::size_t, 0> {};
+
+template <typename T, typename Tail, typename... Ts>
+struct get_index<T, Tail, Ts...> :
+    std::integral_constant<std::size_t, 1 + get_index<T, Ts...>::value> {};
+
+template <typename T>
+struct get_index<T> {
```
Comment (iwrinc) [inline my_variant.h:51]: This probably should have been the implementation of the primary template (line 41)? Anyway, static assert here disables any attempt of SFINAE. Is this intentional?

File: my_variant.h [hunk 36]
```diff
@@ -0,0 +1,118 @@
...
+	my_variant(my_variant<T, types...>&) {
+		assert(sizeof(T) == 0 && "Not implemented");
+	}
+	my_variant(my_variant<T, types...>&&) {
+		assert(sizeof(T) == 0 && "Not implemented");
+	}
+	// need to add emplacement ctors...
+
+	template <class U>
+	my_variant(U&& u) :
```
Comment (iwrinc) [inline my_variant.h:69]: It seems that you disallow the following, don't you: 
`my_variant<size_t> v(1); // please add such a test `

File: my_variant.h [hunk 37]
```diff
@@ -0,0 +1,118 @@
...
+
+//	template <class U> friend U& get<U>(my_variant<T, types...>& my); //compilation issues should be solved
+//private:
+	std::aligned_storage_t<max_size_helper<types...>::value, max_align_helper<types...>::value> m_aligned_storage;
+	void* m_member;
+	size_t m_index;
+};
+
+template <class U, class... types>
+U& get(my_variant<types...>& my) {
```
Comment (iwrinc) [inline my_variant.h:87]: All these templates probably should not be a part of the variant's namespace/interface.

File: my_variant.h [hunk 38]
```diff
@@ -0,0 +1,118 @@
...
+};
+
+/**
+* get index of type from parameter pack
+*/
+
+template <typename T, typename... Ts> struct get_index;
+
+template <typename T, typename... Ts>
+struct get_index<T, T, Ts...> : std::integral_constant<std::size_t, 0> {};
```
Comment (iwrinc) [inline my_variant.h:44]: There is some inconsistency: here you use `integral_constant` (which is fine), and above helpers - manual implementation.

File: derive.h (hunk 26 above)
Comment (alexsolan1) [reply derive.h]: Ok, great, fixed.

File: derive.h (hunk 25 above)
Comment (alexsolan1) [reply derive.h]: Right. Fixed.

File: derive.h (hunk 27 above)
Comment (alexsolan1) [reply derive.h]: OK.

File: derive.h (hunk 28 above)
Comment (alexsolan1) [reply derive.h]: Indeed it seems that there is no  need.

File: derive.h (hunk 30 above)
Comment (alexsolan1) [reply derive.h]: Of course. I thought to convert the -x to +(-1)*x, but I am not sure it worth it.

File: derive.cc (hunk 31 above)
Comment (alexsolan1) [reply derive.cc:24]: I've got a test with zero coeff - line 18. I  will add now another test - more elaborate and a small coeff as well.

Comment (alexsolan1) [reply derive.cc:24]: Added another zero coeff and small coef.

File: derive.h (hunk 29 above)
Comment (alexsolan1) [reply derive.h]: Fixed

File: my_variant.h (hunk 35 above)
Comment (alexsolan1) [reply my_variant.h:51]: So, I am actually not sure what is the right interface. What do we expect for 
get_index<MyClass, double, string, MyOtherClass>::value?

File: my_variant.h (hunk 36 above)
Comment (alexsolan1) [reply my_variant.h:69]: Hmmm, you are right. Will fix it.

Comment (alexsolan1) [reply my_variant.h:69]: Yes, I forget the Head in the max_size, max_align calcs, but now it works.

File: my_variant.h (hunk 38 above)
Comment (alexsolan1) [reply my_variant.h:44]: Since it is an exercise, I played with 2 different styles of doing it. You are right that if it was a production code, the inconsistency is confusing.

File: my_variant.h (hunk 32 above)
Comment (alexsolan1) [reply my_variant.h]: I added a visitor.

File: derive.h [hunk 39]
```diff
@@ -0,0 +1,311 @@
...
+		case plus: os << "+"; break;
+		case minus: os << "-"; break;
+		case mult: os << "*"; break;
+		case power: os << "^"; break;
+		default: assert(0); 
+	}
+	return os;
+}
+
+class Expression {
```
Comment (RotemKalisch) [inline derive.h:30]: Base class with virtual functions should always have a virtual destructor
```
virtual ~Expression() {}
```

File: derive.h [hunk 40]
```diff
@@ -0,0 +1,311 @@
...
+				return add(lhs->derivative(), rhs->derivative());
+			case minus: 
+				return subtract(lhs->derivative(), rhs->derivative());
+			case mult: 
+				return add(
+					multiply(lhs->derivative(), rhs),
+					multiply(rhs->derivative(), lhs)
+				);
+			case power: 
+				{
```
Comment (RotemKalisch) [inline derive.h:150]: Why is this new block?

File: derive.h [hunk 41]
```diff
@@ -0,0 +1,311 @@
...
+	return make_shared_expression(
+		new BinaryExpression{lhs, rhs, power}
+	);
+}
+
+} // End of namespace detail
+
+class Func {
+public:
+	Func(double coef) :
```
Comment (RotemKalisch) [inline derive.h:262]: Convention is:```
Func(...) :
    ....
    ....
{
```

File: derive.h [hunk 42]
```diff
@@ -0,0 +1,311 @@
...
+	{ }	
+
+	Func(char x):
+		m_expr(std::dynamic_pointer_cast<detail::Expression>(
+			shared_ptr<detail::XExpression>(new detail::XExpression{x})
+		))
+	{ }	
+
+	Func(Func&& func) : 
+		m_expr(func.m_expr) {
```
Comment (RotemKalisch) [inline derive.h]:
```
m_expr(std::move(func.m_expr))
```
also then there's no need to do func.m_expr = nullptr

File: my_variant.h (hunk 33 above)
Comment (alexsolan1) [reply my_variant.h]: Fixed.

File: my_variant.h (hunk 34 above)
Comment (alexsolan1) [reply my_variant.h:57]: Yes, but it is more technical now and we were not requested to write  a full solution there.

File: my_variant.h (hunk 37 above)
Comment (alexsolan1) [reply my_variant.h:87]: Right, but since we are not required to do  a full solution here, I will leave it be.

File: my_variant.h [hunk 43]
```diff
@@ -0,0 +1,149 @@
...
+	static_assert(sizeof(T) == 0, "Element is not found");
+};
+
+template <class Obj, class... types> struct destroy_all;
+template <class Head, class... types>
+struct my_variant {
+	my_variant() : m_member(new (&m_aligned_storage) Head{}) { }
+
+	my_variant(my_variant<Head, types...>&) {
+		assert(sizeof(Head) == 0 && "Not implemented");
```
Comment (RotemKalisch) [inline my_variant.h:61]: Are you using and operator on a string? What am I missing?

Comment (alexsolan1) [reply my_variant.h:61]: I am using a limited c assert which does not provide an interface for a message (like static_assert) and this is the patchy way to issue a message. If the lhs condition is not satisfied the rhs is ignored, otherwise the string will be part of the message. It is ugly, but in real life I will have nice library Macros/Methods to it :)

File: my_variant.h (hunk 35 above)
Comment (alexsolan1) [reply my_variant.h:51]: Ok, I understand now. Maybe it is better just to remove it and indeed leave option for SFINAE. Will do.

File: derive.h (hunk 39 above)
Comment (alexsolan1) [reply derive.h:30]: You are right. Added.

File: derive.h (hunk 40 above)
Comment (alexsolan1) [reply derive.h:150]: Because I am using a variable inside and the compiler does not agree to it if it is not in a separate block.

File: derive.h (hunk 42 above)
Comment (alexsolan1) [reply derive.h]: Right! Fixed.
